catalog/registry/catalog-registry-cswinputtransformer/src/main/java/ddf/catalog/registry/transformer/RegistryServiceConverter.java

```diff
@@ -0,0 +1,441 @@
+        String bindingType = null;
+
+        int bindingTypes = 0;
+        while (reader.hasMoreChildren()) {
+            reader.moveDown();
+
+            if ("bindingType".equals(reader.getAttribute("name"))) {
+
+                List<String> types = readValueList(reader);
+
+                if (types.size() != 1) {
+                    throw new ConversionException(
+                            "ServiceBinding with id = " + id + " has invalid bindingType value.");
+                }
+
+                bindingType = types.get(0);
+                bindingTypes += 1;
+            }
+
+            reader.moveUp();
+        }
+
+        if (bindingTypes == 0) {
+            throw new ConversionException(
+                    "ServiceBinding with id = " + id + " has no bindingType.");
+        } else if (bindingTypes > 1) {
+            throw new ConversionException(
+                    "ServiceBinding with id = " + id + " has too many bindingType.");
+        }
+
+        return bindingType;
+    }
+
+    private static final String ID_PATTERN = "urn:uuid:(.*)";
+
+    private static final Pattern PATTERN = Pattern.compile(ID_PATTERN);
+
+    /**
+     * Validate that a service has a correct object type and id.
+     * <p>
+     * NOTE: if the id matches the ID_PATTERN, then it pulls out the uuid.
+     *
+     * @param reader
+     * @return id value
+     * @throws ConversionException
+     */
+    private String validateService(HierarchicalStreamReader reader) throws ConversionException {
+
+        String id = reader.getAttribute("id");
+
+        if (StringUtils.isEmpty(id)) {
+            throw new ConversionException("Service missing id.");
+        }
+
+        Matcher matcher = PATTERN.matcher(id);
+
+        if (matcher.find()) {
+            id = matcher.group(1);
+        }
+
+        String type = reader.getAttribute("objectType");
+
+        if (!"urn:service:catalog".equals(type)) {
+            throw new ConversionException("Service with id = " + id + " bad objectType = " + type);
+        }
+
+        return id;
+    }
+
+    private static Date parseISO8601(String str) {
+        return dateOptionalTimeParser().parseDateTime(str).toDate();
+    }
+
+    private List<Date> readDates(HierarchicalStreamReader reader) {
+        return readValueList(reader).stream().map(RegistryServiceConverter::parseISO8601)
+                .collect(toList());
+    }
+
+    @Override
+    public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext context)
+            throws ConversionException {
+
+        RegistryMetacardImpl meta = new RegistryMetacardImpl(RSMT);
+
+        List<String> bindingTypes = new ArrayList<>();
+        String id = validateService(reader);
+
+        meta.setAttribute(Metacard.ID, id);
+
+        while (reader.hasMoreChildren()) {
+            reader.moveDown();
+
+            switch (reader.getNodeName()) {
+            case "rim:Slot":
+                switch (reader.getAttribute("name")) {
```

Comment:
Instead of nesting the switch cases it'd be better to pull the nested switch into its own method.